File: typescript/tests/core/duplicate-key.test.ts
/**
 * Strict duplicate-key parser tests.
 *
 * The canonical-JSON encoder has always rejected duplicate keys, but the
 * parser used during verification was previously a vanilla `JSON.parse`
 * that silently last-wins on duplicates. A tampered token containing
 * e.g. `{"status":"revoked","status":"active"}` would deserialize as
 * `status=active`. The signature is over the canonical issuer-produced
 * bytes (which can't have duplicates), so this was not exploitable in
 * practice — but the defence-in-depth fix is to reject the duplicate at
 * parse time, before the signature is verified.
 *
 * These tests prove the rejection happens with the right error code on
 * every shape an attacker might try.
 */

import { describe, expect, it } from 'bun:test';
import { encode as b64urlEncode } from '../../src/base64url.ts';
import { decodeUnverified } from '../../src/lic1.ts';
import { StrictJsonError, strictParse } from '../../src/strict-json.ts';

describe('strictParse — duplicate keys', () => {
  it('rejects a top-level duplicate key', () => {
    expect(() => strictParse('{"a":1,"a":2}')).toThrow(StrictJsonError);
    try {
      strictParse('{"a":1,"a":2}');
    } catch (e) {
      expect(e).toBeInstanceOf(StrictJsonError);
      expect((e as StrictJsonError).code).toBe('CanonicalJSONDuplicateKey');
      expect((e as StrictJsonError).message).toContain('duplicate key: a');
    }
  });

  it('rejects a duplicate key inside a nested object', () => {
    try {
      strictParse('{"x":{"y":1,"y":2}}');
      throw new Error('parser did not reject nested duplicate');
    } catch (e) {
      expect(e).toBeInstanceOf(StrictJsonError);
      expect((e as StrictJsonError).code).toBe('CanonicalJSONDuplicateKey');
      expect((e as StrictJsonError).message).toContain('duplicate key: y');
    }
  });

  it('rejects a duplicate key inside an object inside an array', () => {
    try {
      strictParse('{"items":[{"k":1,"k":2}]}');
      throw new Error('parser did not reject array-nested duplicate');
    } catch (e) {
      expect(e).toBeInstanceOf(StrictJsonError);
      expect((e as StrictJsonError).code).toBe('CanonicalJSONDuplicateKey');
    }
  });

  it('accepts repeated values at distinct keys (only key collision matters)', () => {
    const out = strictParse('{"a":"same","b":"same"}') as Record<string, unknown>;
    expect(out.a).toBe('same');
    expect(out.b).toBe('same');
  });

  it('accepts identical key names in sibling scopes (scope is local)', () => {
    const out = strictParse('{"a":{"x":1},"b":{"x":2}}') as Record<string, Record<string, unknown>>;
    expect(out.a?.x).toBe(1);
    expect(out.b?.x).toBe(2);
  });
});

describe('strictParse — round-trip parity with JSON.parse on valid inputs', () => {
  // For every valid (non-duplicate) input, strictParse must produce the
  // same shape JSON.parse does. This is the property that lets us swap it
  // in transparently without breaking any existing fixture.
  const cases: string[] = [
    '{}',
    '{"a":1}',
    '{"unicode":"héllo/"}',
    '{"astral":"😀"}',
    '{"escape":"\\""}',
    '{"int":9007199254740991}',
    '{"neg":-9007199254740991}',
    '{"empty_array":[]}',
    '{"empty_obj":{}}',
    '{"null":null}',
    '{"bool":true,"otherbool":false}',
    '{"deep":{"a":{"b":{"c":{"d":1}}}}}',
    '{"arr":[1,2,3]}',
    '{"mixed":["x",{"y":2},null,true]}',
    '{"frac":1.5}',
    '{"exp":1e10}',
    '{"neg-frac":-0.25}',
  ];
  for (const c of cases) {
    it(`round-trips ${c}`, () => {
      expect(strictParse(c)).toEqual(JSON.parse(c));
    });
  }
});

describe('decodeUnverified — rejects duplicate-key tokens before signature check', () => {
  it('surfaces CanonicalJSONDuplicateKey at parse time, not TokenSignatureInvalid', async () => {
    // Hand-build a wire token with a duplicate-key payload. Sig is junk;
    // the test asserts we never reach signature verification — with the
    // strict parser, the response must be CanonicalJSONDuplicateKey
    // instead of TokenSignatureInvalid.
    const enc = new TextEncoder();
    const header = '{"v":1,"typ":"lic","alg":"ed25519","kid":"x"}';
    const payload = '{"jti":"a","jti":"b"}';
    const headerB64 = b64urlEncode(enc.encode(header));
    const payloadB64 = b64urlEncode(enc.encode(payload));
    const sigB64 = b64urlEncode(enc.encode('not-a-real-signature'));
    const token = `LIC1.${headerB64}.${payloadB64}.${sigB64}`;

    try {
      await decodeUnverified(token);
      throw new Error('decodeUnverified did not reject duplicate-key token');
    } catch (e) {
      expect((e as { code?: string }).code).toBe('CanonicalJSONDuplicateKey');
    }
  });
});

File: licensing/duplicate_key_test.go
package licensing

import (
	"errors"
	"strings"
	"testing"
)

// Strict duplicate-key parser tests.
//
// The canonical-JSON encoder has always rejected duplicate keys, but the
// parser used during verification was previously a vanilla json.Unmarshal
// that silently last-wins on duplicates. A tampered token containing
// e.g. {"status":"revoked","status":"active"} would deserialize as
// status=active. The signature is over the canonical issuer-produced
// bytes (which can't have duplicates), so this was not exploitable in
// practice — but the defence-in-depth fix is to reject the duplicate at
// parse time, before the signature is verified.
//
// These tests prove the rejection happens with the right error code on
// every shape an attacker might try.

func TestParseJSONObject_RejectsDuplicateKeyInPayload(t *testing.T) {
	// Hand-crafted bytes — Go map literals can't carry duplicates.
	raw := []byte(`{"a":1,"a":2}`)
	_, err := parseJSONObject(raw, "payload")
	var le *Error
	if !errors.As(err, &le) || le.Code != CodeCanonicalJSONDuplicateKey {
		t.Fatalf("want CanonicalJSONDuplicateKey, got %v", err)
	}
	if !strings.Contains(le.Message, "duplicate key: a") {
		t.Fatalf("error message should name the offending key: %q", le.Message)
	}
}

func TestParseJSONObject_RejectsDuplicateKeyInHeader(t *testing.T) {
	raw := []byte(`{"v":1,"alg":"ed25519","alg":"hs256","kid":"k","typ":"lic"}`)
	_, err := parseJSONObject(raw, "header")
	var le *Error
	if !errors.As(err, &le) || le.Code != CodeCanonicalJSONDuplicateKey {
		t.Fatalf("want CanonicalJSONDuplicateKey, got %v", err)
	}
	// The duplicated key is alg, not v — message must reflect that so the
	// detection is precise (otherwise we'd be hiding which claim was
	// shadowed).
	if !strings.Contains(le.Message, "duplicate key: alg") {
		t.Fatalf("error message should name the offending key: %q", le.Message)
	}
}

func TestParseJSONObject_RejectsDuplicateKeyInNestedObject(t *testing.T) {
	// Duplicate nested under entitlements — the verifier traverses freely
	// into nested maps so the rejection MUST recurse.
	raw := []byte(`{"jti":"x","entitlements":{"tier":"pro","tier":"free"}}`)
	_, err := parseJSONObject(raw, "payload")
	var le *Error
	if !errors.As(err, &le) || le.Code != CodeCanonicalJSONDuplicateKey {
		t.Fatalf("want CanonicalJSONDuplicateKey, got %v", err)
	}
	if !strings.Contains(le.Message, "duplicate key: tier") {
		t.Fatalf("nested duplicate must be named: %q", le.Message)
	}
}

func TestParseJSONObject_RejectsDuplicateKeyInsideArrayObject(t *testing.T) {
	// Object inside an array also walks through the strict path.
	raw := []byte(`{"items":[{"k":1,"k":2}]}`)
	_, err := parseJSONObject(raw, "payload")
	var le *Error
	if !errors.As(err, &le) || le.Code != CodeCanonicalJSONDuplicateKey {
		t.Fatalf("want CanonicalJSONDuplicateKey, got %v", err)
	}
}

func TestParseJSONObject_AcceptsRepeatedValuesAtDifferentKeys(t *testing.T) {
	// Two distinct keys carrying the same VALUE is not a duplicate; only
	// duplicate KEY names are rejected.
	raw := []byte(`{"a":"same","b":"same"}`)
	obj, err := parseJSONObject(raw, "payload")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if len(obj) != 2 || obj["a"] != "same" || obj["b"] != "same" {
		t.Fatalf("repeated values at different keys must round-trip: %+v", obj)
	}
}

func TestParseJSONObject_AcceptsSameKeyInSiblingScopes(t *testing.T) {
	// "x" appears in two sibling sub-objects — perfectly valid because
	// each sub-object has its own scope. The duplicate guard MUST be
	// scope-local, not global.
	raw := []byte(`{"a":{"x":1},"b":{"x":2}}`)
	obj, err := parseJSONObject(raw, "payload")
	if err != nil {
		t.Fatalf("sibling scopes with same-named keys must parse: %v", err)
	}
	if len(obj) != 2 {
		t.Fatalf("expected 2 top-level keys, got %d", len(obj))
	}
}

// TestVerify_RejectsDuplicateKeyTokenBeforeSigCheck proves the rejection
// happens at parse time, before the cryptographic verification step.
// This is what makes the fix defence-in-depth: even if an attacker
// somehow produced a valid signature over duplicate-key bytes, the parser
// stops the token before the signature is checked.
func TestVerify_RejectsDuplicateKeyTokenBeforeSigCheck(t *testing.T) {
	// Hand-build a wire token with a duplicate-key payload. We don't sign
	// it correctly — the test asserts that the parser fails BEFORE the
	// signature check, which would fail with TokenSignatureInvalid if the
	// parser ran first. With the strict parser, the response must be
	// CanonicalJSONDuplicateKey instead.
	header := `{"v":1,"typ":"lic","alg":"ed25519","kid":"x"}`
	payload := `{"jti":"a","jti":"b"}`
	headerB64 := Base64urlEncode([]byte(header))
	payloadB64 := Base64urlEncode([]byte(payload))
	// Sig is junk; the test asserts we never reach signature verification.
	sigB64 := Base64urlEncode([]byte("not-a-real-signature"))
	token := "LIC1." + headerB64 + "." + payloadB64 + "." + sigB64

	_, err := DecodeUnverified(token)
	var le *Error
	if !errors.As(err, &le) {
		t.Fatalf("expected *Error, got %T %v", err, err)
	}
	if le.Code != CodeCanonicalJSONDuplicateKey {
		t.Fatalf("strict parser must surface CanonicalJSONDuplicateKey; got %s (msg=%s)",
			le.Code, le.Message)
	}
}
